### PR TITLE
FILTER: Align Geometries

### PR DESCRIPTION
**Pull request by mmarineBlueQuartz**
* Added AlignGeometries filter
* Added unit test

### CODE REVIEW

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/AlignGeometries.cpp [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+
+#include <string>
+
+namespace
+{
+constexpr complex::int32 k_EMPTY_PARAMETER = -123;
+
+complex::FloatVec3 extractOrigin(const complex::AbstractGeometry* geometry)
+{
+  if(auto* image = dynamic_cast<const complex::ImageGeom*>(geometry))
```
Comment (JDuffeyBQ) [inline src/Plugins/ComplexCore/src/ComplexCore/Filters/AlignGeometries.cpp:23]: These `if` blocks could be replaced with a `switch` on `AbstractGeometry::Type`. And the same for the other functions.

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/AlignGeometries.cpp [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+#include "complex/Parameters/GeometrySelectionParameter.hpp"
+#include "complex/Utilities/DataArrayUtilities.hpp"
+
+#include <string>
+
+namespace
+{
+constexpr complex::int32 k_EMPTY_PARAMETER = -123;
+
+complex::FloatVec3 extractOrigin(const complex::AbstractGeometry* geometry)
```
Comment (JDuffeyBQ) [inline src/Plugins/ComplexCore/src/ComplexCore/Filters/AlignGeometries.cpp:21]:
```suggestion
complex::FloatVec3 extractOrigin(const complex::AbstractGeometry& geometry)
```

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/AlignGeometries.cpp [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+    centroid[0] /= static_cast<float>(geometry3d->getNumberOfVertices());
+    centroid[1] /= static_cast<float>(geometry3d->getNumberOfVertices());
+    centroid[2] /= static_cast<float>(geometry3d->getNumberOfVertices());
+    return centroid;
+    ;
+  }
+  return centroid;
+}
+
+void translateGeometry(complex::AbstractGeometry* geometry, const complex::FloatVec3& translation)
```
Comment (JDuffeyBQ) [inline src/Plugins/ComplexCore/src/ComplexCore/Filters/AlignGeometries.cpp:259]:
```suggestion
void translateGeometry(complex::AbstractGeometry& geometry, const complex::FloatVec3& translation)
```

**Review by JDuffeyBQ — CHANGES_REQUESTED** [review]

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/AlignGeometries.cpp [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+}
+
+std::string AlignGeometries::humanName() const
+{
+  return "Align Geometries";
+}
+
+Parameters AlignGeometries::parameters() const
+{
+  GeometrySelectionParameter::AllowedTypes geomTypes{AbstractGeometry::Type::Any};
```
[the remark this answers is not in the document]
Comment (mmarineBlueQuartz) [reply src/Plugins/ComplexCore/src/ComplexCore/Filters/AlignGeometries.cpp:404]: Unit tests will not pass using the empty set. It says the geometry is not the correct type.

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/AlignGeometries.cpp [hunk 5]
```diff
@@ -0,0 +1,471 @@
...
+          origin[j] = vertices[3 * i + j];
+        }
+      }
+    }
+    return origin;
+  }
+  break;
+  // 2D
+  case complex::AbstractGeometry::Type::Triangle:
+    [[falthrough]] case complex::AbstractGeometry::Type::Quad:
```
Comment (JDuffeyBQ) [inline src/Plugins/ComplexCore/src/ComplexCore/Filters/AlignGeometries.cpp:98]:
```suggestion
    [[fallthrough]];
  case complex::AbstractGeometry::Type::Quad:
```
There's a missing "l" in the attribute. And I'd add a semicolon after the attribute to put it on it's own line. I think it makes it more readable, and it's how the example on cppreference is.

**Review by JDuffeyBQ — CHANGES_REQUESTED** [review]

**Comment by JDuffeyBQ** [comment]
The other `[[fallthrough]]` attributes need to fixed as well, but after that this PR should be good to go.

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/AlignGeometries.cpp [hunk 6]
```diff
@@ -0,0 +1,471 @@
...
+    {
+      vertices[3 * i + 0] += translation[0];
+      vertices[3 * i + 1] += translation[1];
+      vertices[3 * i + 2] += translation[2];
+    }
+    return;
+  }
+    // 2D Geometries
+  case complex::AbstractGeometry::Type::Quad:
+    [[falthrough]] case complex::AbstractGeometry::Type::Triangle:
```
Comment (JDuffeyBQ) [inline src/Plugins/ComplexCore/src/ComplexCore/Filters/AlignGeometries.cpp:348]:
```suggestion
    [[fallthrough]] case complex::AbstractGeometry::Type::Triangle:
```

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/AlignGeometries.cpp [hunk 7]
```diff
@@ -0,0 +1,471 @@
...
+      {
+        vertices[3 * i + 0] += translation[0];
+        vertices[3 * i + 1] += translation[1];
+        vertices[3 * i + 2] += translation[2];
+      }
+      return;
+    }
+    // 3D Geometries
+  case complex::AbstractGeometry::Type::Hexahedral:
+    [[falthrough]] case complex::AbstractGeometry::Type::Tetrahedral:
```
Comment (JDuffeyBQ) [inline src/Plugins/ComplexCore/src/ComplexCore/Filters/AlignGeometries.cpp:362]:
```suggestion
    [[fallthrough]] case complex::AbstractGeometry::Type::Tetrahedral:
```

**Review by JDuffeyBQ — CHANGES_REQUESTED** [review]

**Review by JDuffeyBQ — APPROVED** [review]

**Pull request merged**